agent/csi/libstorage/csi_service_libstorage.go

```diff
@@ -192,25 +204,72 @@ func (d *driver) ControllerPublishVolume(
 			"invalid volume capability request"), nil
 	}
 
-	iid, err := toInstanceID(req.NodeId)
+	_, err := toInstanceID(req.NodeId)
 	if err != nil {
 		return gocsi.ErrControllerPublishVolume(
 			csi.Error_ControllerPublishVolumeError_INVALID_NODE_ID,
 			err.Error()), nil
 	}
-	_ = iid
 
-	//d.client.Storage().VolumeAttach(d.ctx, volumeID, opts)
+	opts := &apitypes.VolumeAttachOpts{Opts: apiutils.NewStore()}
 
-	return nil, nil
+	// Get next device via Executor
+	nd, err := d.client.Executor().NextDevice(d.ctx, apiutils.NewStore())
```

Comment:
Hi @codenrhoden,

Not all storage drivers require the next device logic, and anyway, it's not needed at all since the libStorage `libstorage` storage driver that is used by the libStorage client handles that for you:

```go
func (d *driver) VolumeAttach(
	ctx types.Context,
	volumeID string,
	opts *types.VolumeAttachOpts) (*types.Volume, string, error) {

	if d.isController() {
		return nil, "", utils.NewUnsupportedForClientTypeError(
			d.clientType, "VolumeAttach")
	}

	ctx = d.requireCtx(ctx)
	serviceName, ok := context.ServiceName(ctx)
	if !ok {
		return nil, "", goof.New("missing service name")
	}

	nextDevice, err := d.NextDevice(ctx, utils.NewStore())
	if err != nil {
		return nil, "", err
	}

	var nextDevicePtr *string
	if nextDevice != "" {
		nextDevicePtr = &nextDevice
	}

	req := &types.VolumeAttachRequest{
		NextDeviceName: nextDevicePtr,
		Force:          opts.Force,
		Opts:           opts.Opts.Map(),
	}

	return d.client.VolumeAttach(ctx, serviceName, volumeID, req)
}
```

All you should need to do is use `Storage().VolumeAttach`.